### PR TITLE
Fixes for java.util.collection bugs.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Will need rebase on #2085.

**Comment by sjrd** [comment]
This can be rebased, now. I'll wait for the rebase before reviewing so that the did view is manageable.

**Comment by nicolasstucki** [comment]
Rebased.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala [hunk 1]
```diff
@@ -183,8 +182,11 @@ trait MapTest {
     mp.put("ONE", "one")
     assertTrue(mp.containsKey("ONE"))
     assertFalse(mp.containsKey("TWO"))
-    if (!executingInJVM) // Issue #2078
+    if (factory.allowsNullKeysQueries)
       assertFalse(mp.containsKey(null))
+    else
+      expectThrows(classOf[Throwable], mp.containsKey(null))
+
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/MapTest.scala]: Spurious blank line.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnListsTest.scala [hunk 2]
```diff
@@ -7,97 +7,110 @@ import org.junit.Test
...
+
+  def factory: ListFactory
+
+  @Test def sort_on_comparables(): Unit = {
+    if (factory.sortableUsingCollections) {
+      test_sort_on_comparables[CustomComparable](factory,
+          new CustomComparable(_), false)
+      test_sort_on_comparables[jl.Integer](factory, jl.Integer.valueOf)
+      test_sort_on_comparables[jl.Long](factory, _.toLong)
+      test_sort_on_comparables[jl.Double](factory, _.toDouble)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnListsTest.scala]: Doesn't `CollectionsOnCopyOnWriteArrayListTestOnJDK8` perform exactly the same 4 calls (and same thing in `sort_with_comparator`)? Couldn't the 4 calls be abstracted away too? If not, this requires a `// !!!` comment on both sides to highlight the code duplication.

File: javalib/src/main/scala/java/util/concurrent/CopyOnWriteArrayList.scala [hunk 3]
```diff
@@ -150,7 +150,7 @@ class CopyOnWriteArrayList[E <: AnyRef] private (private var inner: js.Array[E])
   }
 
   def addAllAbsent(c: Collection[_ <: E]): Int = {
-    val elemsToAdd = c.iterator().filter(!contains(_)).toList
+    val elemsToAdd = c.iterator().filter(!contains(_)).toSet
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/concurrent/CopyOnWriteArrayList.scala]: This is not correct. `addAllAbsent` guarantees that new elements will be added in the order returned by `c.iterator()`. The `toSet` would destroy that ordering.

What about a much simpler implementation altogether:
```scala
for (e <- c.iterator())
  addIfAbsent(e)
```
?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsTest.scala [hunk 4]
```diff
@@ -14,20 +14,50 @@ import org.junit.Test
...
+    expectThrows(classOf[UnsupportedOperationException], coll.add(elem))
+    expectThrows(classOf[UnsupportedOperationException], coll.addAll(List(elem)))
+    expectThrows(classOf[Exception], coll.remove(elem))
+    expectThrows(classOf[Exception], coll.removeAll(List(elem)))
+    expectThrows(classOf[Exception], coll.retainAll(List(elem)))
+    expectThrows(classOf[Throwable], coll.clear())
+  }
+
+  private def checkImmutablilityOfCollectionApi[E](set: ju.Set[E],
+      elem: E): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsTest.scala]: I'm always wary of overloads where a parameter has a supertype of another overload. Consider renaming this to `checkImmutabilityOfSetApi`. Same remark for the other two overloads below.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashMapTest.scala [hunk 5]
```diff
@@ -10,27 +10,32 @@ package org.scalajs.testsuite.javalib.util
...
-class LinkedHashMapInsertionOrderTest extends LinkedHashMapTest {
-  override def factory: LinkedHashMapFactory = new LinkedHashMapFactory(false, false)
+class LinkedHashMapInsertionOrderTest extends LinkedHashMapTest
+
+class LinkedHashMapInsertionOrderLimitedTest extends LinkedHashMapTest {
+  override def factory: LinkedHashMapFactory =
+    new LinkedHashMapFactory(accessOrder = false, withSizeLimit = Some(50))
 }
 
+
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashMapTest.scala]: Double blank line.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashMapTest.scala [hunk 6]
```diff
@@ -39,17 +44,13 @@ abstract class LinkedHashMapTest extends HashMapTest {
...
     (0 until 100).foreach(key => lhm.put(key, s"elem $key"))
 
-    def expectedKey(index: Int): Int = {
-      if (withSizeLimit)
-        50 + index
-      else
-        index
-    }
+    def expectedKey(index: Int): Int =
+      withSizeLimit.sum + index
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashMapTest.scala]: Lol! Calling `sum` on an `Option`!
Consider `.getOrElse(0)` instead, which I believe is much more readable, not to mention close to the intent.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashMapTest.scala [hunk 7]
```diff
@@ -184,15 +174,14 @@ abstract class LinkedHashMapTest extends HashMapTest {
         }
       } else {
         // accesses shouldn't modify the order
-        if (withSizeLimit) 50 + index
-        else index
+        withSizeLimit.sum + index
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/LinkedHashMapTest.scala]: Same here.

**Comment by sjrd** [comment]
That's all.

File: javalib/src/main/scala/java/util/concurrent/CopyOnWriteArrayList.scala (hunk 3 above)
Comment (nicolasstucki) [inline javalib/src/main/scala/java/util/concurrent/CopyOnWriteArrayList.scala]: Yes, that should work.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnListsTest.scala (hunk 2 above)
Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsOnListsTest.scala]: Yes, it can be factored out.

**Comment by nicolasstucki** [comment]
Updated.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsTest.scala [hunk 8]
```diff
@@ -14,20 +14,47 @@ import org.junit.Test
...
+    checkImmutablilityOfCollectionApi(set, elem)
+
+  private def checkImmutablilityOfListApi[E](list: ju.List[E], elem: E): Unit = {
+    checkImmutablilityOfCollectionApi(list, elem)
+    expectThrows(classOf[UnsupportedOperationException], list.add(0, elem))
+    expectThrows(classOf[UnsupportedOperationException], list.addAll(0, List(elem)))
+    expectThrows(classOf[UnsupportedOperationException], list.remove(0))
+  }
+
+
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/CollectionsTest.scala]: Double blank line.

**Comment by sjrd** [comment]
That's all.

**Comment by nicolasstucki** [comment]
Updated.

**Comment by sjrd** [comment]
LGTM

**Comment by sjrd** [comment]
retest this please

**Pull request merged**